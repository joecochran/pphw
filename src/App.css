/* BASE */

html {
  height: 100%;
  box-sizing: border-box;
  font-family: 'Gilroy';
}
*, *:before, *:after {
  box-sizing: inherit;
}

body {
  height: 100%;
  background-color: #E4E6E8;
}

#root {
  height: 100%;
}

.App {
  height: 100%;
}

.layout {
  display: flex;
  flex-direction: column;
  height: 100vh;
}
.row-1 {
  flex-shrink: 0;
}
.row-2 {
  flex-grow: 1;
  height: 100%;
  display: flex;
}

/* HEADER */

.c-header {
  background-color: #FFF;
  border-bottom: 1px solid #C1C4C7;
  display: flex;
  height: 48px;
}

.c-header__brand {
  align-items: center;
  border-right: 1px solid #C1C4C7;
  display: flex;
  flex-basis: 152px;
  flex-shrink: 0;
  justify-content: center;
  width: 152px;
}

.c-header__logo {
  max-width: 100%;
  width: 120px;
  transform: translate3d(0,0,0);
}

.c-header__foo {
  align-items: center;
  color: #2D3439;
  display: flex;
  font-family: Helvetica;
  font-size: 16px;
  justify-content: space-between;
  line-height: 22px;
  padding: 0 9px 0 16px;
  width: 100%;
}

/* SEARCH */

.c-search__trigger {
  align-items: center;
  background-color: transparent;
  border: 0;
  display: flex;
  height: 32px;
  justify-content: center;
  padding: 0;
  width: 32px;
}

/* NAV */

.c-nav {
  align-items: flex-end;
  background: #fff;
  border-bottom: 1px solid #C1C4C7;
  display: flex;
  height: 80px;
  justify-content: space-between;
  padding: 0 206px 0 32px;
}

.c-nav__title {
  color: #222428;
  font-family: 'Gilroy';
  font-weight: bold;
  font-size: 32px;
  line-height: 40px;
  margin: 0 0 8px;
  
}

.c-nav__list {
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c-nav__item {
  padding: 0 11px 6px;
}

.c-nav__item.is-active {
  box-shadow: 0 6px 0 #1973BA;
}

.c-nav__link {
  color: #5E6266;
  display: block;
  font-family: 'Gilroy';
  font-size: 16px;
  line-height: 22px;
  padding: 0;
  text-decoration: none;
}

.o-main {
  display: flex;
  width: 100%;
  flex-shrink: 1;
  
}

.c-roadmap {
  flex-grow: 1;
  padding: 0 32px;
  overflow-x: scroll;
}

.c-sidebar {
  width: 185px;
  flex-shrink: 0;
  background-color: #FFF;
  padding: 16px;
  border-left: 1px solid #C1C4C7;
}

.c-draggable {
  padding: 11px 8px;
  background-color: #E4E6E8;
  margin-bottom: 8px;
  border-radius: 4px;
  font-family: Helvetica;
  font-size: 14px;
  color: #5E6266;
  line-height: 20px;
  cursor: grab;
}

.c-draggable__icon {
  margin-right: 8px;
}

/* Timeline */

.c-timeline {
  display: grid;
  grid-template-columns: repeat(13, 100px);
  list-style: none;
  margin: 0;
  padding: 17px 0 0 18px;
  margin-bottom: 16px;
}

.c-timeline__month {
  background-image: url('data:image/svg+xml;charset=US-ASCII,%3Csvg%20width%3D%224%22%20height%3D%224%22%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%3E%0A%20%20%3Crect%20y%3D%2224%22%20width%3D%224%22%20height%3D%224%22%20rx%3D%222%22%20transform%3D%22translate%280%20-24%29%22%20fill%3D%22%23C1C4C7%22%20fill-rule%3D%22nonzero%22/%3E%0A%3C/svg%3E%0A');
  background-repeat: no-repeat;
  height: 28px;
  background-position: bottom left;
}

.c-lane-target {
  background: rgba(255,255,255,0.80);
  border: 1px dashed #C1C4C7;
  height: 93px;
  font-family: Gilroy;
  font-weight: bold;
  font-size: 20px;
  color: #2D3439;
  text-align: center;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 16px;
}
.c-lane {
  width: 1320px;
  margin-bottom: 16px;
}



.c-lane__header {
  background-color: #EF6031;
  padding: 8px;
  border-radius: 3px 3px 0 0;
  color: #fff;
  user-select: none;
  cursor: pointer;
  display: flex;
  align-items: center;
}

.c-lane__header > .i-arrow {
  height: 6px;
  width: 10px;
}

.is-collapsed .i-arrow {
  transform: rotate(-90deg)
}

.is-collapsed .c-lane__header {
  border-radius: 3px;
}

.c-lane__header > span {
  margin-left: 8px;
}

.c-lane__body {
  background: repeating-linear-gradient(90deg, #E0E4E7 0,  #E0E4E7 1px, #FFF 1px, #FFF 100px);
  background-position: 18px;
  min-height: 54px;
  border: 1px solid #C1C4C7;
  border-top: 0;
  border-radius: 0 0 3px 3px;
  position: relative;
}

.is-collapsed > .c-lane__body {
  display: none;
}

.c-lane__body--row2 {
  min-height: 100px;
}

.c-lane__bar {
  background: #fff;
  height: 38px;
  width: 383px;
  border-radius: 4px;
  border: 1px dashed #C1C4C7;
  position: absolute;
  top: 8px;
  left: 8px;
}
.c-lane__bar:nth-of-type(2) {
  left: 248px;
  top: 54px;
}
.c-lane__bar:nth-of-type(3) {
  left: 418px;
}

.c-lane__bar--filled {
  background-color: #1973BA;
  border: 1px solid #1973BA;
  pointer-events: none;
  display: flex;
  align-items: center;
  padding-left: 16px;
  font-family: Helvetica;
  font-size: 16px;
  color: #FFFFFF;
  line-height: 22px;
}
