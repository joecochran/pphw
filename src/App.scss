/* BASE */

html {
  height: 100%;
  box-sizing: border-box;
  font-family: 'Gilroy', Helvetica, Arial, sans-serif;
}

*, *:before, *:after {
  box-sizing: inherit;
}

body {
  height: 100%;
  background-color: #E4E6E8;
}

#root {
  height: 100%;
}

.App {
  height: 100%;
}

.layout {
  display: flex;
  flex-direction: column;
  height: 100vh;
}

.row-1 {
  flex-shrink: 0;
}

.row-2 {
  flex-grow: 1;
  height: 100%;
  display: flex;
}

.o-main {
  display: flex;
  width: 100%;
  flex-shrink: 1;
}

.c-roadmap {
  flex-grow: 1;
  padding: 0 32px;
  overflow-x: scroll;
}
