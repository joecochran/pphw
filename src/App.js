import React, { Component } from 'react';
import Header from './components/Header/Header';
import Nav from './components/Nav/Nav';
import Sidebar from './components/Sidebar/Sidebar';
import Timeline from './components/Timeline/Timeline';
import LaneTarget from './components/LaneTarget/LaneTarget';
import Lane from './components/Lane/Lane';

import './App.scss';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showLaneDropTarget: false,
      lanes: [],
      step: 0,
    }
  }

  componentDidMount() {
    setTimeout(() => {
      this.setState({step: 1});
    }, 333);
  }

  handleLaneDrag = (event) => {
    event.dataTransfer.dropEffect = "move";
    event.dataTransfer.setData('text/plain', 'lane');
    this.setState({showLaneDropTarget: true});
  }

  handleBarDrag = (event) => {
    event.dataTransfer.dropEffect = "move";
    event.dataTransfer.setData('text/plain', 'bar');
  }

  addLane = (event) => {
    event.preventDefault();
    const target = event.dataTransfer.getData("text/plain");
    if (target === "lane") {
      this.setState((currentState) => ({
        lanes: currentState.lanes.concat({id: this.state.lanes.length + 1}),
        showLaneDropTarget: false,
      }));
    }
    if (this.state.step === 2) {
      setTimeout(() => {
        this.setState({step: 3})
      }, 333);
    }
  }

  handleDragEnd = (event) => {
    event.preventDefault();
    this.setState({showLaneDropTarget: false});
  }

  updateStep = () => {
    this.setState((currentState) => ({
      step: currentState.step + 1
    }));
  }

  render() {
    return (
      <div className="layout">
        <div className="row-1">
          <Header /> 
          <Nav />
        </div>
        <div className="row-2">
          <main className="o-main">
            <div className="c-roadmap">
              <Timeline />
              {this.state.lanes.map(lane => (
                <Lane id={lane.id} key={lane.id} handleStepUpdate={this.updateStep} />
              ))}
              {this.state.showLaneDropTarget && (
                <LaneTarget handleLaneDrop={this.addLane} />
              )}
            </div>
            <Sidebar handleLaneDrag={this.handleLaneDrag} handleDragEnd={this.handleDragEnd} handleBarDrag={this.handleBarDrag} step={this.state.step} handleStepUpdate={this.updateStep} />
          </main>
        </div>
      </div>
    );
  }
}

export default App;
