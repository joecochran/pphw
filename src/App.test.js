import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';

const wrapper = shallow(<App />);
const eventMock = {
  preventDefault() {},
  dataTransfer: {
    getData() {
      return 'lane';
    },
    setData() {}
  }
}

describe('<App>', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('does delayed step update', () => {
    expect(wrapper.state('step')).toBe(0);
  });

  it('shows lane drop target', () => {
    wrapper.instance().handleLaneDrag(eventMock);
    expect(wrapper.state('showLaneDropTarget')).toBe(true);
  });

  it('only adds a lane with correct data', () => {
    eventMock.dataTransfer.getData = () => 'bar';
    wrapper.instance().addLane(eventMock);
    expect(wrapper.state('lanes').length).toBe(0);
    eventMock.dataTransfer.getData = () => 'lane';
    wrapper.instance().addLane(eventMock);
    expect(wrapper.state('lanes').length).toBe(1);
  });

  it('allows bar drag', () => {
    wrapper.instance().handleBarDrag(eventMock);
    expect(wrapper).toMatchSnapshot();
  });

  it('hides drop target on drag end', () => {
    wrapper.instance().handleDragEnd(eventMock);
    expect(wrapper.state('showLaneDropTarget')).toBe(false);
  });


})



