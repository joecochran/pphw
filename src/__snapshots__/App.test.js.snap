// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> allows bar drag 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="row-1"
        >
          <Header />
          <Header />
        </div>,
        <div
          className="row-2"
        >
          <main
            className="o-main"
          >
            <div
              className="c-roadmap"
            >
              <Timeline />
              <Lane
                handleStepUpdate={[Function]}
                id={1}
              />
            </div>
            <Sidebar
              handleBarDrag={[Function]}
              handleDragEnd={[Function]}
              handleLaneDrag={[Function]}
              handleStepUpdate={[Function]}
              step={0}
            />
          </main>
        </div>,
      ],
      "className": "layout",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Header />,
            <Header />,
          ],
          "className": "row-1",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <main
            className="o-main"
          >
            <div
              className="c-roadmap"
            >
              <Timeline />
              <Lane
                handleStepUpdate={[Function]}
                id={1}
              />
            </div>
            <Sidebar
              handleBarDrag={[Function]}
              handleDragEnd={[Function]}
              handleLaneDrag={[Function]}
              handleStepUpdate={[Function]}
              step={0}
            />
          </main>,
          "className": "row-2",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="c-roadmap"
              >
                <Timeline />
                <Lane
                  handleStepUpdate={[Function]}
                  id={1}
                />
              </div>,
              <Sidebar
                handleBarDrag={[Function]}
                handleDragEnd={[Function]}
                handleLaneDrag={[Function]}
                handleStepUpdate={[Function]}
                step={0}
              />,
            ],
            "className": "o-main",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Timeline />,
                  Array [
                    <Lane
                      handleStepUpdate={[Function]}
                      id={1}
                    />,
                  ],
                  false,
                ],
                "className": "c-roadmap",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "class",
                  "props": Object {
                    "handleStepUpdate": [Function],
                    "id": 1,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleBarDrag": [Function],
                "handleDragEnd": [Function],
                "handleLaneDrag": [Function],
                "handleStepUpdate": [Function],
                "step": 0,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "main",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row-1"
          >
            <Header />
            <Header />
          </div>,
          <div
            className="row-2"
          >
            <main
              className="o-main"
            >
              <div
                className="c-roadmap"
              >
                <Timeline />
                <Lane
                  handleStepUpdate={[Function]}
                  id={1}
                />
              </div>
              <Sidebar
                handleBarDrag={[Function]}
                handleDragEnd={[Function]}
                handleLaneDrag={[Function]}
                handleStepUpdate={[Function]}
                step={0}
              />
            </main>
          </div>,
        ],
        "className": "layout",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Header />,
              <Header />,
            ],
            "className": "row-1",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <main
              className="o-main"
            >
              <div
                className="c-roadmap"
              >
                <Timeline />
                <Lane
                  handleStepUpdate={[Function]}
                  id={1}
                />
              </div>
              <Sidebar
                handleBarDrag={[Function]}
                handleDragEnd={[Function]}
                handleLaneDrag={[Function]}
                handleStepUpdate={[Function]}
                step={0}
              />
            </main>,
            "className": "row-2",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="c-roadmap"
                >
                  <Timeline />
                  <Lane
                    handleStepUpdate={[Function]}
                    id={1}
                  />
                </div>,
                <Sidebar
                  handleBarDrag={[Function]}
                  handleDragEnd={[Function]}
                  handleLaneDrag={[Function]}
                  handleStepUpdate={[Function]}
                  step={0}
                />,
              ],
              "className": "o-main",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Timeline />,
                    Array [
                      <Lane
                        handleStepUpdate={[Function]}
                        id={1}
                      />,
                    ],
                    false,
                  ],
                  "className": "c-roadmap",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {},
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": "1",
                    "nodeType": "class",
                    "props": Object {
                      "handleStepUpdate": [Function],
                      "id": 1,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  false,
                ],
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "handleBarDrag": [Function],
                  "handleDragEnd": [Function],
                  "handleLaneDrag": [Function],
                  "handleStepUpdate": [Function],
                  "step": 0,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "main",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
